fedimint-core/src/encoding/mod.rs

```diff
@@ -716,4 +716,67 @@ mod tests {
                 .is_err()
         );
     }
+
+    fn encode_value<T: Encodable>(value: &T) -> Vec<u8> {
+        let mut writer = Vec::new();
+        value.consensus_encode(&mut writer).unwrap();
+        writer
+    }
+
+    fn decode_value<T: Decodable>(bytes: &Vec<u8>) -> T {
+        T::consensus_decode(&mut Cursor::new(bytes), &ModuleDecoderRegistry::default()).unwrap()
+    }
+
+    fn keeps_ordering_after_serialization<T: Ord + Encodable + Decodable + Debug>(mut vec: Vec<T>) {
+        vec.sort();
+        let mut encoded = vec.iter().map(encode_value).collect::<Vec<_>>();
+        encoded.sort();
+        let decoded = encoded.iter().map(decode_value).collect::<Vec<_>>();
+        for (i, (a, b)) in vec.iter().zip(decoded.iter()).enumerate() {
+            assert_eq!(a, b, "difference at index {i}");
+        }
+    }
+
+    #[test]
+    fn test_lexicographical_sorting() {
+        #[derive(Ord, PartialOrd, Eq, PartialEq, Debug, Encodable, Decodable)]
+        struct TestAmount(u64);
+        let amounts = (0..20000).map(TestAmount).collect::<Vec<_>>();
+        keeps_ordering_after_serialization(amounts);
+
+        #[derive(Ord, PartialOrd, Eq, PartialEq, Debug, Encodable, Decodable)]
+        struct TestComplexAmount(u16, u32, u64);
+        let complex_amounts = (10..20000)
+            .flat_map(|i| {
+                (i - 1..=i + 1).flat_map(move |j| {
+                    (i - 1..=i + 1).map(move |k| TestComplexAmount(i as u16, j as u32, k as u64))
+                })
+            })
+            .collect::<Vec<_>>();
+        keeps_ordering_after_serialization(complex_amounts);
+
+        #[derive(Ord, PartialOrd, Eq, PartialEq, Debug, Encodable, Decodable)]
+        struct Text(String);
+        let texts = (' '..'~')
+            .flat_map(|i| {
+                (' '..'~')
+                    .map(|j| Text(format!("{i}{j}")))
+                    .collect::<Vec<_>>()
+            })
+            .collect::<Vec<_>>();
+        keeps_ordering_after_serialization(texts);
+
+        // FIXME: bitcoin structures are not lexicographically sortable,
+        // perhaps we can use a marker trait to indicate this fact
+        // #[derive(Ord, PartialOrd, Eq,
+        // PartialEq, Debug, Encodable, Decodable)]
+        // struct TestNonce(secp256k1_zkp::XOnlyPublicKey);
+        // let nonces = (0..1000)
+        //     .map(|_| {
+        //         let (_, pk) = secp256k1_zkp::generate_keypair(&mut
+        // rand::rngs::OsRng);         TestNonce(pk.x_only_public_key().
+        // 0)     })
+        //     .collect::<Vec<_>>();
+        // keeps_ordering_after_serialization(nonces);
```

Comment:
Not sure if ordering is a real problem for nonces, they don't really have a semantic ordering, unlike amounts. But having the same ordering in rust using `Ord` and of the serialized value might be good indeed ... Not sure a marker trait would help much though.

Reply:
Probably the problem is just hypothetical right know, but eventually it may cause a bug. If `Nonce` isn't of a concern perhaps `TransactionId` or another bitcoin structure is.

Reply:
Bitocoin's hashes are a mess one way or the other. Encoded one way, displayed in reverse. The exact expectation of the person writing the code about the ordering can't be assumed.

We could probably have some newtype adapter that wraps a `Encodable` and encodes it in reverse (or forces LE/BE explicitly) for cases like that, but probably not necessary right now. A test that serializes a block and tx hash from `rust-bitcoin` and then makes sure we've sticked to *something* should be enough.

Reply:
I've added some serialization of rust bitcoin structures